Bottom: 3155b9b25baf3c5f0c94431c566a23537cdab5ea
Top:    dd8baaf567c556625ad7ac4041c1f8dc7cb52ed0
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-06-01 16:32:12 +0800

Refresh of tun-ex

---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 1585aa7..b0a7667 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -425,7 +425,14 @@ static void handle_tx(struct vhost_net *net)
 			nvq->upend_idx = (nvq->upend_idx + 1) % UIO_MAXIOV;
 		} else {
 			struct skb_msg m;
-			m.flags = 0;
+			if (!vhost_vq_avail_empty(&net->dev, vq) &&
+				vq->delayed <= 4) {
+				vq->delayed ++;
+				m.flags |= SKB_MSG_MORE;
+			} else {
+				vq->delayed = 0;
+				m.flags = 0;
+			}
 			msg.msg_control = &m;
 			msg.msg_controllen = sizeof(msg);
 			ubufs = NULL;
diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 669fef1..4d2a20b 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -304,6 +304,7 @@ static void vhost_vq_reset(struct vhost_dev *dev,
 	vhost_reset_is_le(vq);
 	vhost_disable_cross_endian(vq);
 	vq->busyloop_timeout = 0;
+	vq->delayed = 0;
 }
 
 static int vhost_worker(void *data)
diff --git a/drivers/vhost/vhost.h b/drivers/vhost/vhost.h
index d36d8be..9bd3a15 100644
--- a/drivers/vhost/vhost.h
+++ b/drivers/vhost/vhost.h
@@ -116,6 +116,7 @@ struct vhost_virtqueue {
 	bool user_be;
 #endif
 	u32 busyloop_timeout;
+        u32 delayed;
 };
 
 struct vhost_dev {
